### ISSUE TITLE
fix: investigate and address RAM accumulation across Ray worker processes in long sweeps

### ISSUE THREAD

**Issue by xiaohan2012** [xiaohan2012/stgym#140]
## Observation

During a full design-dimension sweep (`sweep-all-20260405`, experiment 8, ~2,900 runs), system RAM grows monotonically from **9 GB at sweep start to 81 GB after 2.5 hours** — entirely from completed runs, not the current workload.

### Memory growth timeline

Sampled from MLflow `system/system_memory_usage_megabytes`:

```
16:51   9 GB  ← sweep start (OS + Ray + MLflow baseline)
17:00  20 GB
17:30  33 GB
17:48  50 GB  ← first mouse-kidney loads
18:00  36 GB  ← MK workers finish, partially freed
18:30  43 GB
19:03  60 GB
19:18  77 GB
19:26  81 GB  ← only 15 GB free; first OOM kill occurs here
19:31  57 GB  ← OOM kill frees ~24 GB
```

The floor never returns to baseline. After the first MK episode ends at 18:00, the floor sits at ~36 GB (was 9 GB at start) and climbs to 77 GB over the next 1.5 hours purely from small-dataset runs (brca, brca-grade, spatial-vdj etc.).

### Impact

This accumulation is what ultimately causes mouse-kidney OOM kills in production sweeps — not just the instantaneous peak of a single load. By the time a mouse-kidney worker acquires the gate and tries to load (~29 GB needed), the background RSS has consumed all headroom. See PR #136 for details.

### Possible causes

**1. Python heap fragmentation** — `pymalloc` arenas (256 KB) are only released to the OS when *every* object in them is freed. Each dataset load creates thousands of small Python objects (PyG `Data` fields, dict entries, TensorBase metadata). After large tensors are freed via `mmap`, small objects scattered across arenas pin those pages in RSS indefinitely.

**2. Genuine memory leak** — The ~27 GB floor rise over 1.5 hours of small-dataset-only runs is large for pure fragmentation. Candidate sources:
- PyTorch Lightning trainer or callbacks retaining a reference to the data module
- MLflow logger holding a model or data reference after the run ends
- Circular references in PyG `Data` objects not yet collected by Python's cycle GC

### Suggested investigation

1. Add `gc.collect()` after each `run_exp` call. If RSS drops → fragmentation. If not → genuine leak.
2. Profile worker RSS before and after a single `run_exp` call using `psutil.Process().memory_info().rss`.
3. Use `tracemalloc` or `objgraph` inside a worker to identify which object types accumulate across tasks.

### Suggested fixes (once root cause confirmed)

- **`ray.remote(max_calls=N)`** — restart the worker process after N tasks, returning all heap to the OS. Simple and guaranteed to work.
- **`gc.collect()` + `ctypes.CDLL("libc.so.6").malloc_trim(0)`** between tasks — aggressively returns unused heap pages without restarting the process.
- **Fix the leak** (if genuine) — identify and break the reference cycle keeping dataset objects alive after `run_exp` returns.

**Comment by xiaohan2012** [comment]
## Verification of diagnosis against codebase

**Confirmed correct:**
- Driver is small (~667 MB), workers hold the RAM — matches the code. `run_rct.py` only dispatches tasks; all heavy work (dataset load, model, trainer) lives in the Ray worker running `run_exp`.
- No `max_calls` is set anywhere in the current code (`run_rct.py:72`), so Ray reuses workers indefinitely. `max_calls=1` is the right primary fix.
- `torch.cuda.empty_cache()` is never called between trials — real gap.

**Inapplicable to this codebase:**
- `plt.close('all')` — no matplotlib in the `run_exp` code path.
- Module-level dataset caches (`_DATASET_CACHE` etc.) — none exist in `stgym/`.
- DataLoader subprocess leaks — `num_workers=0` is the default (`config_schema.py:241`), so no persistent subprocess workers.

**Worth scrutinizing:**
- **MLflow logger** — no explicit `mlflow.end_run()` call. The logger's run ends only when the logger object is GC'd. If `pl.Trainer` holds a reference to the logger (it does — passed via `logger=logger`), the run and any attached state may not be released until the trainer is collected.
- **PL Trainer ↔ data module cycle** — `trainer.fit(model, datamodule=datamodule)` and `trainer.test(model, datamodule=datamodule)` (`train.py:79-80`) store internal references. The trainer going out of scope in `train()` doesn't guarantee immediate collection if there's a reference cycle.

## Proposed fixes (ranked by impact)

1. **`max_calls=1`** (or small N like 5) passed to `.options()` in `run_rct.py` — guaranteed process-level cleanup, ~5–15 s restart overhead per trial is negligible vs. multi-minute runs.
2. **`gc.collect()` + `torch.cuda.empty_cache()`** at end of `run_exp` — cheap, handles CUDA allocator cache and Python heap fragmentation.
3. **Explicit `del trainer, model_module, data_module`** before `return True` in `run_exp` — breaks PL/MLflow reference cycles sooner without waiting for GC.

**Comment by xiaohan2012** [comment]
## Diagnostic instrumentation added (branch `feat/mem-diag`)

Added RSS telemetry + `gc.collect()` + `objgraph` to `run_exp` to distinguish fragmentation from genuine leaks. See PR branch for details.

### What was added to `stgym/rct/run.py`

At the **start** of each `run_exp` call:
```python
_proc = psutil.Process(os.getpid())
_rss_start_gb = _proc.memory_info().rss / 1e9
logz_logger.info(f"[mem-diag] PID={_proc.pid} rss_start={_rss_start_gb:.2f} GB")
objgraph.show_growth(limit=10)   # object types that grew since last call on this worker
```

At the **end** (before `return True`):
```python
gc.collect()
_rss_after_gc_gb = _proc.memory_info().rss / 1e9
logz_logger.info(f"[mem-diag] PID={_proc.pid} rss_after_gc={_rss_after_gc_gb:.2f} GB  delta={_rss_after_gc_gb - _rss_start_gb:+.2f} GB")
objgraph.show_growth(limit=10)   # what's still live after gc
```

### Interpretation guide

| Pattern | Conclusion |
|---|---|
| `rss_start` grows each task on the same PID | Retained objects between calls (leak or fragmentation) |
| `rss_after_gc` significantly lower than `rss_end` | Fragmentation — gc releases pymalloc arenas |
| `rss_after_gc` stays high despite gc | Genuine object leak — check `objgraph` output for accumulating types |

### Local test results (inconclusive)

Ran a short debug sweep (test datasets, CPU-only). RSS values stayed under 0.4 GB — test datasets are too small to exhibit the production pattern. `rss_start` was not monotonically increasing at this scale, and `gc.collect()` freed memory in one case (−0.09 GB) but not others.

**Next step**: run on cyy2 with the full sweep and grep Ray worker logs for `[mem-diag]` to get a real signal.

**Comment by xiaohan2012** [comment]
## Production sweep confirmation (sweep-all-20260405-164959, exp 8)

The latest full sweep provides concrete evidence that RAM accumulation is what halts production sweeps.

### Throughput collapse

| Snapshot (local time) | Sweep duration | Completed last 2h | Total ok / failed |
|---|---|---|---|
| Apr 7 ~08:00 | 43.7 h | 124 | 3430 / 192 |
| Apr 8 ~11:00 | 53.0 h | 23 | 4584 / 267 |
| Apr 8 ~22:00 | identical | identical | identical |
| Apr 9 ~07:00 | identical | identical | identical |

Throughput dropped from ~124 runs/2h to 23 and then zero. The status was completely frozen for at least 20 hours until the user manually terminated the sweep.

### OOM kills confirmed in Ray logs

Three Ray sessions were spawned on Apr 8, each immediately encountering OOM pressure — evidence that background RSS was already saturated before any new session could make progress.

| Ray session | OOM kills | Time range (UTC) |
|---|---|---|
| session_2026-04-08_06-27-04 | 1 | 06:58 |
| session_2026-04-08_08-33-44 | 7 | 09:39 – 10:29 |
| session_2026-04-08_10-56-56 | 8 | 11:49 – 15:20 |

**16 OOM kills total on Apr 8.** The last kill at 15:20 UTC triggered a complete halt — no new trials were dispatched after that point.

### Stale RUNNING runs after halt

149 MLflow runs were left in `RUNNING` state (never marked FAILED) across all 10 started dimensions:

| dimension | stale active |
|---|---|
| model.dim_inner | 30 |
| model.pooling.n_clusters | 27 |
| train.max_epoch | 23 |
| train.optim.base_lr | 19 |
| data_loader.knn_k | 6 |
| model.pooling.type | 13 |
| model.post_mp_dims | 11 |
| model.use_batchnorm | 12 |
| train.optim.optimizer | 7 |
| data_loader.batch_size | 1 |

This is the stale-run problem from issue #107 — workers killed by the OOM monitor leave runs in RUNNING forever.

### Unexecuted dimensions

3 dimensions never started due to the halt:
- `data_loader.radius_ratio` (300 trials)
- `model.act` (300 trials)
- `model.num_mp_layers` (300 trials)

~900 trials lost.

### Memory recovery after termination

After the user manually killed the sweep, RAM returned to ~92 GB free of 96 GB total — confirming the accumulation was entirely from live processes, not kernel page cache.

### Root cause assessment

The three Ray session restarts (all immediately OOM-killed) point to background RSS being the blocker, not any single large allocation. Each session started with the system already near exhaustion from accumulated worker heap. This is consistent with the leak/fragmentation hypothesis in the issue body — the floor never returns to baseline, so by ~53h into a sweep, there is no headroom left even for small-dataset tasks.

**Comment by xiaohan2012** [comment]
## Triage

**Type:** bug | **Priority:** P1 | **Status:** status/ready
**Blockers:** none
**Assessment:** RAM grows from 9GB to 81GB over 2.5h of sweep execution, independent of workload. Blocks reliable long-running sweeps; needs profiling of Ray worker lifecycle and MLflow caching.

**Comment by xiaohan2012** [comment]
## Priority upgrade: P1 -> P0

Re-triaged in the context of the user's goal: get `./scripts/design-dimension-sweep-all.sh` to complete end-to-end on cyy2.

**Rationale:** Production sweep `sweep-all-20260405-164959` (exp 8) confirmed this is the dominant blocker. The RAM floor climbs from ~9 GB to 81 GB over 2.5h purely from completed runs, then throughput collapses after ~43h (comment in this issue shows identical totals across three snapshots on Apr 8-9). This means the full sweep literally cannot finish without addressing worker RAM accumulation — it's not a tail-latency concern, it halts the sweep.

Every other sweep reliability issue (#133, #138, #98, #107) is downstream of this: once RAM is exhausted the kernel SIGKILLs workers, leaving stale MLflow runs and claimed Ray GPU slots.

Bumping to P0 so it's the top priority on the critical path to E2E sweep completion.

**Comment by xiaohan2012** [comment]
## Diagnostic Run Results (mem-diag-140 experiment) — Final

Ran a diagnostic sweep on cyy2 using the instrumentation from PR #143. Analysis follows the [interpretation guide](https://github.com/xiaohan2012/stgym/issues/140#issuecomment-4196412296).

### Experiment Summary
- **Config:** `+exp=batch_size design_space=graph_clf resource=gpu-6 sample_size=50`
- **Total runs:** 336 (324 finished, 12 failed)
- **Workers:** 6
- **Duration:** ~4 hours

---

## Pattern 1: `rss_start` grows each task (leak or fragmentation indicator)

| Worker PID | Tasks | First `rss_start` | Last `rss_start` (floor) | Growth | Per-task |
|------------|-------|-------------------|--------------------------|--------|----------|
| 2262667 | 22 | 0.84 GB | 2.79 GB | +1.95 GB | 91 MB |
| 2262670 | 42 | 0.84 GB | 2.48 GB | +1.64 GB | 40 MB |
| 2262671 | 39 | 0.84 GB | 3.11 GB | +2.27 GB | 60 MB |
| 2262674 | 34 | 0.84 GB | 3.11 GB | +2.27 GB | 68 MB |
| 2262680 | 30 | 0.84 GB | 3.42 GB | +2.58 GB | 88 MB |
| 2262682 | 33 | 0.84 GB | 3.19 GB | +2.35 GB | 73 MB |
| **Avg** | **33** | **0.84 GB** | **3.02 GB** | **+2.18 GB** | **~67 MB** |

**Conclusion:** YES — `rss_start` grows monotonically → retained objects between tasks

---

## Pattern 2: `rss_after_gc` drops after large loads (fragmentation vs leak)

| Worker PID | gc calls | Large recoveries (>5 GB freed) | Avg recovery | Small accumulations |
|------------|----------|--------------------------------|--------------|---------------------|
| 2262667 | 22 | 2 | -8.65 GB | 6 (+0.52 GB avg) |
| 2262670 | 42 | 4 | -8.30 GB | 14 (+0.38 GB avg) |
| 2262671 | 39 | 5 | -8.38 GB | 10 (+0.51 GB avg) |
| 2262674 | 34 | 5 | -8.42 GB | 11 (+0.63 GB avg) |
| 2262680 | 30 | 4 | -7.98 GB | 9 (+0.85 GB avg) |
| 2262682 | 33 | 3 | -8.48 GB | 10 (+0.63 GB avg) |
| **Total** | **200** | **23** | **-8.34 GB** | **60** |

**Conclusion:** YES — `gc.collect()` recovers ~8 GB from large dataset spikes → **FRAGMENTATION, not a hard leak**

---

## Pattern 3: `rss_after_gc` stays high despite gc (genuine leak indicator)

| Worker PID | First 3 `rss_after_gc` | Last 3 `rss_after_gc` | Trend |
|------------|------------------------|----------------------|-------|
| 2262667 | 2.12, 2.03, 2.03 GB | 3.17, 2.79, 2.92 GB | +0.80 GB |
| 2262670 | 2.34, 2.04, 2.30 GB | 3.87, 2.46, 2.65 GB | +0.31 GB |
| 2262671 | 2.21, 2.27, 2.35 GB | 2.85, 3.32, 3.11 GB | +0.90 GB |
| 2262674 | 2.10, 2.65, 3.10 GB | 3.13, 4.19, 3.11 GB | +1.01 GB |
| 2262680 | 2.73, 4.44, 4.13 GB | 3.28, 3.42, 3.27 GB | +0.54 GB |
| 2262682 | 2.13, 2.37, 2.37 GB | 3.32, 3.19, 3.03 GB | +0.90 GB |

**Conclusion:** PARTIALLY — floor rises ~0.5-1 GB over ~30 tasks, but large spikes always recover. **Mild fragmentation accumulation, NOT a hard leak.**

---

## Diagnosis

Based on the interpretation guide:

| Pattern | Observed? | Interpretation |
|---------|-----------|----------------|
| `rss_start` grows each task | ✅ YES (+67 MB/task) | Retained objects |
| `rss_after_gc` drops after large loads | ✅ YES (-8.34 GB avg) | gc.collect() effective → **fragmentation** |
| `rss_after_gc` stays high despite gc | ⚠️ PARTIAL (floor rises) | Not a hard leak, but arenas not returned |

**Root cause: Heap fragmentation** — pymalloc arenas are not returned to the OS even after `gc.collect()` frees the objects within them.

---

## Extrapolation to Production (2,900-run sweep)

| Metric | Calculation | Result |
|--------|-------------|--------|
| Tasks per worker | 2900 ÷ 6 | ~483 |
| Floor per worker | 0.84 + (483 × 0.067) | **~33 GB** |
| Aggregate floor | 6 × 33 GB | **~200 GB** |
| Available RAM | — | 96 GB |

This explains why production sweeps halt after ~40-50 hours.

---

## Recommended Fix

```python
run_exp_remote = ray.remote(run_exp).options(
    num_cpus=res_cfg.num_cpus_per_trial,
    num_gpus=gpu_allocation,
    max_calls=50,  # Recycle worker after 50 tasks
)
```

With `max_calls=50`:
- Max floor per worker: 0.84 + (50 × 0.067) = **~4.2 GB**
- Aggregate: 6 × 4.2 = **~25 GB** — safe headroom for long sweeps

---

*See PR #143 comment for system memory plots (before/after comparison).*

**Comment by xiaohan2012** [comment]
## Diagnostic Results - Test 3 Complete ✅

All three diagnostic tests are now complete. Here's the summary:

### Test Matrix Results

| Test | Configuration | Runs | Memory Trend (first 20 → last 20 avg) |
|------|---------------|------|---------------------------------------|
| 1 | `gc.collect()` enabled | 324 | 19.4 → 29.1 GB (**+9.6 GB**) |
| 2 | `gc.collect()` disabled | 331 | 14.8 → 28.1 GB (**+13.3 GB**) |
| 3 | `max_calls=1` | 324 | 22.2 → 14.6 GB (**-7.6 GB**) |

### Key Findings

1. **Root Cause Confirmed: Heap Fragmentation**
   - `gc.collect()` only marginally helps (9.6 GB vs 13.3 GB growth)
   - This confirms the issue is **not** reference cycles/memory leaks
   - It's pymalloc arena fragmentation (memory allocated but not returned to OS)

2. **Solution Validated: `max_calls=1`**
   - Memory actually *decreases* over time (-7.6 GB)
   - Workers are recycled after each task, preventing fragmentation buildup
   - Zero stale runs (0 vs 12 in Test 1)

3. **Trade-off Analysis**
   - `max_calls=1`: Maximum memory efficiency, slight overhead from worker respawn
   - `max_calls=N` (N > 1): Balance between efficiency and spawn overhead
   - Recommended: Start with `max_calls=1`, tune upward if spawn overhead is noticeable

### Visualization

﻿
<img width="1784" height="1475" alt="Image" src="https://github.com/user-attachments/assets/a543d902-9cab-48c0-93fe-40a48160df49" />


### Next Steps

- [x] Confirm root cause (heap fragmentation)
- [x] Validate `max_calls` fix
- [ ] Merge PR #143 with `STGYM_MAX_CALLS` env var support
- [ ] Document recommended settings in README/CLAUDE.md
